File: native-core/src/box2d/ChainShape.cpp
#include "ChainShape.h"
#include "Shape.h"
#include "Collision/Shapes/b2ChainShape.h"

ChainShape_M(jlong, newChainShape) (JNIEnv *, jclass)
{
	return (jlong)(new b2ChainShape());
}

ChainShape_M(void, jniCreateLoop) (JNIEnv *env, jobject obj, jfloatArray vals, jint offset, jint numVertices)
{
	jfloat *data = (jfloat *) env->GetPrimitiveArrayCritical(vals, 0);
	b2Vec2* verticesOut = new b2Vec2[numVertices];
	for( int i = 0; i < numVertices; i++ )
		verticesOut[i] = b2Vec2(data[offset+(i<<1)], data[offset+(i<<1)+1]);
	((b2ChainShape*)env->GetLongField(obj, shapePtr))->CreateLoop( verticesOut, numVertices );
	delete[] verticesOut;
	env->ReleasePrimitiveArrayCritical(vals, data, 0);
}
	
ChainShape_M(void, jniCreateChain) (JNIEnv *env, jobject obj, jfloatArray vals, jint offset, jint numVertices)
{
	jfloat *data = (jfloat *) env->GetPrimitiveArrayCritical(vals, 0);
	b2Vec2* verticesOut = new b2Vec2[numVertices];
	for( int i = 0; i < numVertices; i++ )
		verticesOut[i] = b2Vec2(data[offset+(i<<1)], data[offset+(i<<1)+1]);
	((b2ChainShape*)env->GetLongField(obj, shapePtr))->CreateChain( verticesOut, numVertices );
	delete[] verticesOut;
	env->ReleasePrimitiveArrayCritical(vals, data, 0);
}
ChainShape_M(void, setPrevVertex) (JNIEnv *env, jobject obj, jfloat x, jfloat y)
{
	((b2ChainShape*)env->GetLongField(obj, shapePtr))->SetPrevVertex(b2Vec2(x, y));
}
ChainShape_M(void, setNextVertex) (JNIEnv *env, jobject obj, jfloat x, jfloat y)
{
	((b2ChainShape*)env->GetLongField(obj, shapePtr))->SetNextVertex(b2Vec2(x, y));
}
ChainShape_M(jint, getVertexCount) (JNIEnv *env, jobject obj)
{
	return ((b2ChainShape*)env->GetLongField(obj, shapePtr))->GetVertexCount();
}

ChainShape_M(void, jniGetVertex) (JNIEnv *env, jobject obj, jint index, jfloatArray vals)
{
	jfloat *data = (jfloat *) env->GetPrimitiveArrayCritical(vals, 0);
	const b2Vec2 v = ((b2ChainShape*)env->GetLongField(obj, shapePtr))->GetVertex( index );
	data[0] = v.x;
	data[1] = v.y;
	env->ReleasePrimitiveArrayCritical(vals, data, 0);
}

File: core/src/main/java/com/ariasaproject/advancerofrpg/physics/box2d/ChainShape.java
package com.ariasaproject.advancerofrpg.physics.box2d;

import com.ariasaproject.advancerofrpg.math.Vector2;

public class ChainShape extends Shape {
    private static float[] verts = new float[2];
    boolean isLooped = false;

    public ChainShape() {
        addr = newChainShape();
    }

    ChainShape(long addr) {
        this.addr = addr;
    }

    private native long newChainShape();

    @Override
    public Type getType() {
        return Type.Chain;
    }

    public void createLoop(float[] vertices) {
        jniCreateLoop(vertices, 0, vertices.length / 2);
        isLooped = true;
    }

    public void createLoop(float[] vertices, int offset, int length) {
        jniCreateLoop(vertices, offset, length / 2);
        isLooped = true;
    }

    public void createLoop(Vector2[] vertices) {
        float[] verts = new float[vertices.length * 2];
        for (int i = 0, j = 0; i < vertices.length * 2; i += 2, j++) {
            verts[i] = vertices[j].x;
            verts[i + 1] = vertices[j].y;
        }
        jniCreateLoop(verts, 0, verts.length / 2);
        isLooped = true;
    }

    private native void jniCreateLoop(float[] verts, int offset, int numVertices);

    public void createChain(float[] vertices) {
        jniCreateChain(vertices, 0, vertices.length / 2);
        isLooped = false;
    }

    public void createChain(float[] vertices, int offset, int length) {
        jniCreateChain(vertices, offset, length / 2);
        isLooped = false;
    }

    public void createChain(Vector2[] vertices) {
        float[] verts = new float[vertices.length * 2];
        for (int i = 0, j = 0; i < vertices.length * 2; i += 2, j++) {
            verts[i] = vertices[j].x;
            verts[i + 1] = vertices[j].y;
        }
        jniCreateChain(verts, 0, vertices.length);
        isLooped = false;
    }

    private native void jniCreateChain(float[] verts, int offset, int numVertices);

    public void setPrevVertex(Vector2 prevVertex) {
        setPrevVertex(prevVertex.x, prevVertex.y);
    }

    public native void setPrevVertex(float x, float y);

    public void setNextVertex(Vector2 nextVertex) {
        setNextVertex(nextVertex.x, nextVertex.y);
    }

    public native void setNextVertex(float x, float y);

    public native int getVertexCount();

    public void getVertex(int index, Vector2 vertex) {
        jniGetVertex(index, verts);
        vertex.x = verts[0];
        vertex.y = verts[1];
    }

    private native void jniGetVertex(int index, float[] verts);

    public boolean isLooped() {
        return isLooped;
    }

// /// Implement b2Shape. Vertices are cloned using b2Alloc.
// b2Shape* Clone(b2BlockAllocator* allocator) const;
//
// /// @see b2Shape::GetChildCount
// int32_t GetChildCount() const;
//
// /// Get a child edge.
// void GetChildEdge(b2EdgeShape* edge, int32_t index) const;
//
// /// This always return false.
// /// @see b2Shape::TestPoint
// bool TestPoint(const b2Transform& transform, const b2Vec2& p) const;
//
// /// Implement b2Shape.
// bool RayCast(b2RayCastOutput* output, const b2RayCastInput& input,
// const b2Transform& transform, int32_t childIndex) const;
//
// /// @see b2Shape::ComputeAABB
// void ComputeAABB(b2AABB* aabb, const b2Transform& transform, int32_t childIndex) const;
//
// /// Chains have zero mass.
// /// @see b2Shape::ComputeMass
// void ComputeMass(b2MassData* massData, float32 density) const;
//
}
